flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchProcessor.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.elasticsearch.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.functions.Function;
+import org.apache.flink.api.connector.sink.SinkWriter;
+
+import org.elasticsearch.action.ActionRequest;
+
+import java.io.Serializable;
+
+/**
+ * Creates multiple {@link ActionRequest ActionRequests} from the incoming elements.
+ *
+ * <p>This is used by sinks to prepare elements for sending them to Elasticsearch.
+ *
+ * <p>Example:
+ *
+ * <pre>{@code
+ * private static class TestElasticsearchProcessor implements ElasticsearchProcessor<Tuple2<Integer, String>> {
+ *
+ *     public IndexRequest createIndexRequest(Tuple2<Integer, String> element) {
+ *         Map<String, Object> json = new HashMap<>();
+ * 		   json.put("data", element.f1);
+ *
+ * 	       return Requests.indexRequest()
+ * 		       .index("my-index")
+ * 			   .type("my-type")
+ * 			   .id(element.f0.toString())
+ * 			   .source(json);
+ *     }
+ *
+ * 	   public void process(Tuple2<Integer, String> element, RequestIndexer indexer) {
+ * 	       indexer.add(createIndexRequest(element));
+ *     }
+ * }
+ *
+ * }</pre>
+ *
+ * @param <T> The type of the element handled by this {@link ElasticsearchProcessor}
+ */
+@PublicEvolving
+public interface ElasticsearchProcessor<T> extends Function, Serializable {
+
+    /**
+     * Initialization method for the function. It is called once before the actual working process
+     * methods.
+     */
+    default void open() throws Exception {}
+
+    /** Tear-down method for the function. It is called when the sink closes. */
+    default void close() throws Exception {}
+
+    /**
+     * Process the incoming element to produce multiple {@link ActionRequest ActionRequests}. The
+     * produced requests should be added to the provided {@link RequestIndexer}.
+     *
+     * @param element incoming element to process
+     * @param context to access additional information about the record
+     * @param indexer request indexer that {@code ActionRequest} should be added to
+     */
+    void process(T element, SinkWriter.Context context, RequestIndexer indexer);
```

Comment:
Looking at Elasticsearch's APIs in the past I think it is a good idea to not make any ES API part of Flink's public API to easier support multiple versions. 

I also feel that `Requests` or `BatchRequests` are not really correct here because it implies that you always add to a single request. The class was called `RequestIndexer` before so I have no hard feeling about finding a good replacement.

Reply:
I'm wondering if we can't simply return a list of requests?

Reply:
If we return a list of requests we loose the type information and would need to do the respective casts later, isn't this performance critical?
I understand what you are aiming for to make it more applicable to the other serialization schemas it would definitely help to have a return type.